### PR TITLE
BUG/API: tighter checks on DTI/TDI.equals

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #33531
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Also fixes `TimedeltaIndex(Categorical(tdi))` (which DatetimeIndex already gets right)

### CODE REVIEW

File: pandas/core/indexes/datetimelike.py [hunk 1]
```diff
@@ -137,14 +138,30 @@ def equals(self, other: object) -> bool:
...
+                "timedelta",
+                "timedelta64",
+                "datetime",
+                "datetime64",
+                "date",
+                "period",
+            ]
+
+            should_try = False
+            if other.dtype == object:
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py:151]: not is_dtype_equal?

**Review by jreback — CHANGES_REQUESTED** [review]
minor comment, ping on green.

File: pandas/core/indexes/datetimelike.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py:151]: i tend to avoid that for perf

**Comment by jreback** [comment]
ci / checks are failing

**Comment by jbrockmendel** [comment]
mypy complaints fixed; unrelated windows+plotting failure

**Review by jreback — APPROVED** [review]

**Pull request merged**